Src/IronPython.Modules/_codecs.cs

```diff
@@ -68,146 +72,70 @@ public static class PythonCodecs {
         }
 
         /// <summary>
-        /// Optimied encoding mapping that can be consumed by charmap_encode.
+        /// Encodes the input string with the specified optimized encoding map.
         /// </summary>
-        [PythonHidden]
-        public class EncodingMap {
-            internal Dictionary<int, char> Mapping = new Dictionary<int, char>();
+        public static PythonTuple charmap_encode(CodeContext context, [BytesConversion]string input, string errors, [NotNull]EncodingMap map) {
+            return CharmapDecodeWorker(context, input, errors, new EncodingMapEncoding(map, errors), false);
         }
 
-        /// <summary>
-        /// Decodes the input string using the provided string mapping.
-        /// </summary>
-        public static PythonTuple charmap_decode([BytesConversion]string input, string errors, [NotNull]string map) {
-            StringBuilder res = new StringBuilder();
-            for (int i = 0; i < input.Length; i++) {
-                var charIndex = (int)input[i];
-
-                if (map.Length <= charIndex) {
-                    if (errors == "strict") {
-                        throw PythonOps.UnicodeDecodeError("failed to find key in mapping");
-                    }
-                    res.Append("\ufffd");
-                } else {
-                    res.Append(map[input[i]]);
-                }
-            }
-            return PythonTuple.MakeTuple(res.ToString(), res.Length);
+        public static object charmap_encode(CodeContext context, [BytesConversion]string input, string errors = "strict", IDictionary<object, object> map = null) {
```

Comment:
I don't think the python side cares, but if you think its a good idea, I'll do it